src/code/z_camera.c

```diff
@@ -4351,100 +4352,131 @@ s32 Camera_Subj4(Camera* camera) {
+    if (rwData->zoomTimer != 0) {
+        atEyeTargetOffset.r = 10.0f;
+        atEyeTargetOffset.pitch = 0x238C; // ~50 degrees
+        atEyeTargetOffset.yaw = rwData->yaw;
+        Camera_Vec3fVecSphGeoAdd(&zoomAtTarget, &playerPosRot.pos, &atEyeTargetOffset);
+
+        vZoomTimer = rwData->zoomTimer + 1.0f;
+        at->x = F32_LERPIMPINV(at->x, zoomAtTarget.x, vZoomTimer);
+        at->y = F32_LERPIMPINV(at->y, zoomAtTarget.y, vZoomTimer);
+        at->z = F32_LERPIMPINV(at->z, zoomAtTarget.z, vZoomTimer);
+
+        atEyeOffset.r -= (atEyeOffset.r / vZoomTimer);
+        atEyeOffset.yaw = BINANG_LERPIMPINV(atEyeOffset.yaw, (s16)(playerPosRot.rot.y - 0x7FFF), rwData->zoomTimer);
+        atEyeOffset.pitch = BINANG_LERPIMPINV(atEyeOffset.pitch, playerPosRot.rot.x, rwData->zoomTimer);
+        Camera_Vec3fVecSphGeoAdd(eyeNext, at, &atEyeOffset);
         *eye = *eyeNext;
-        rwData->unk_32--;
+        rwData->zoomTimer--;
         return false;
-    } else if (rwData->unk_24 < 0.5f) {
+    } else if (rwData->xzSpeed < 0.5f) {
         return false;
     }
 
-    Actor_GetWorldPosShapeRot(&sp6C, &camera->player->actor);
-    Math3D_LineClosestToPoint(&rwData->unk_00, &sp6C.pos, eyeNext);
-    at->x = eyeNext->x + rwData->unk_00.dir.x;
-    at->y = eyeNext->y + rwData->unk_00.dir.y;
-    at->z = eyeNext->z + rwData->unk_00.dir.z;
+    Actor_GetWorldPosShapeRot(&playerPosRot, &camera->player->actor);
+    Math3D_LineClosestToPoint(&rwData->crawlspaceLine, &playerPosRot.pos, eyeNext);
+
+    // *at is unused before getting overwritten later this function
+    at->x = eyeNext->x + rwData->crawlspaceLine.dir.x;
+    at->y = eyeNext->y + rwData->crawlspaceLine.dir.y;
+    at->z = eyeNext->z + rwData->crawlspaceLine.dir.z;
+
     *eye = *eyeNext;
-    sp64.yaw = rwData->unk_30;
-    sp64.r = 5.0f;
-    sp64.pitch = 0x238C;
-    Camera_Vec3fVecSphGeoAdd(&sp98, eyeNext, &sp64);
-    rwData->unk_2C += 0xBB8;
-    temp_f16 = Math_CosS(rwData->unk_2C);
-    eye->x += (sp98.x - eye->x) * fabsf(temp_f16);
-    eye->y += (sp98.y - eye->y) * fabsf(temp_f16);
-    eye->z += (sp98.z - eye->z) * fabsf(temp_f16);
-
-    if ((rwData->unk_28 < temp_f16) && !rwData->unk_2E) {
+
+    atEyeTargetOffset.yaw = rwData->yaw;
+    atEyeTargetOffset.r = 5.0f;
+    atEyeTargetOffset.pitch = 0x238C; // ~50 degrees
+
+    Camera_Vec3fVecSphGeoAdd(&vEyeTarget, eyeNext, &atEyeTargetOffset);
+    rwData->lineOffsetAngle += 0xBB8;
+    lineOffset = Math_CosS(rwData->lineOffsetAngle);
+
+    // VEC3F_LERPIMPDST(eye, eye, &sp98, fabsf(temp_f16))
+    eye->x += (vEyeTarget.x - eye->x) * fabsf(lineOffset);
+    eye->y += (vEyeTarget.y - eye->y) * fabsf(lineOffset);
+    eye->z += (vEyeTarget.z - eye->z) * fabsf(lineOffset);
+
+    // When camera reaches the peak of offset and starts to move down
+    // && alternating cycles (sfx plays only every 2nd cycle)
+    if ((rwData->lineOffsetPrev < lineOffset) && !rwData->isSfxOff) {
```

Comment:
Doesn't `lineOffsetPrev < lineOffset` check for lineOffset starting to go up (at its lowest) instead of going down (at a peak)?

I'm not sure if `atEyeTargetOffset.pitch = 0x238C` is upwards or downwards otherwise I could tell what I'd expect from the vertical camera movement

Reply:
so `atEyeTargetOffset.pitch = 0x238C` is a VecSphGeo so points upwards
so vEyeTarget is above eyeNext
so with eyeLerp=0 eye=eyeNext is lower than vEyeTarget
and with eyeLerp=+-1 eye=vEyeTarget is higher than eyeNext

`if ((eyeLerp > rwData->eyeLerp) && !rwData->isSfxOff) {`
passes when eyeLerp starts raising, i.e. raising from -1
i.e. indeed at a peak height of the eye pos
the "alternating" (every 2 up/down) is because eye=vEyeTarget at eyeLerp=1 too

TLDR it's okay